internal/credentials/kubernetes/ecr/access_key.go

```diff
@@ -23,118 +21,101 @@ const (
 	secretKey = "awsSecretAccessKey"
 )
 
-type accessKeyCredentialHelper struct {
+type AccessKeyProvider struct {
 	tokenCache *cache.Cache
 
-	// The following behaviors are overridable for testing purposes:
-
-	getAuthTokenFn func(
-		ctx context.Context,
-		region string,
-		accessKeyID string,
-		secretAccessKey string,
-	) (string, error)
+	getAuthTokenFn func(ctx context.Context, region, accessKeyID, secretAccessKey string) (string, error)
 }
 
-// NewAccessKeyCredentialHelper returns an implementation of credentials.Helper
-// that utilizes a cache to avoid unnecessary calls to AWS.
-func NewAccessKeyCredentialHelper() credentials.Helper {
-	a := &accessKeyCredentialHelper{
+func NewAccessKeyProvider() *AccessKeyProvider {
```

Comment:
Applies for all providers... would it be more conventional to return `credentials.Provider` here? If we did so, individual provider implementations also need not be exported.

Reply:
:thinking: I always had interpreted that as `return <concrete type>` and not so much that the return type in the signature should be concrete. I do think it produces cleaner abstractions where callers don't need to concern themselves _at all_ with the underlying implementation.

Reply:
> it should be noted that this is not a hard and fast rule. There are times when it makes sense to accept concrete types and return interfaces, or to accept interfaces and return concrete types.

I won't die on this hill, but only want to establish that I'm not nuts.

> If we did so, individual provider implementations also need not be exported.

If we _start_ there with "callers have no need to be aware of the underlying implementation, so I'm not exporting it," then returning an interface makes sense.

However, if you have defensible reasons for the implementations to be exported, then my scenario does not apply.

Either way, not a blocker.